### PR TITLE
Avoiding indent after index entries in latex

### PR DESCRIPTION
**Pull request by aszekMosek**
Subject: Remove unnecessary indent after an index entry in latex

### Feature or Bugfix

- Bugfix

### Purpose

When index entries are added in the middle of a paragraph but not inline, for example between a math equation and following text like here:

```
\end{equation}
\index{domain}\index{feasible set}
and we call \eqref{equation:linear:eq-lo-lo1} an 
```

then the first word (in this case *and*) is preceded in pdf by a small indent of the size of roughly one space.

### Detail

This fix solves the problem; for instance the above is typeset as

```
\end{equation}
\index{domain}%
\index{feasible set}%

and we call \eqref{equation:linear:eq-lo-lo1} an 
```

Now the *and* is aligned to the left as it should be.

### Relates
- <URL or Ticket>

### CODE REVIEW

**Comment by jfbu** [comment]
Thanks for the report and bugfix. Can you please rebase it to stable HEAD? I would like to merge this for 1.6.4 release after reviewing. I will add a test, too (I was surprised your PR passed the tests as it changes latex writer output).

For the record, this is an issue with the index _directive_, as in

```rst
This is normal text, now an equation.

.. math::

   E = m c^2

.. index:: Einstein, relativity

Einstein's equation is famous.
And some more text.
```

If one uses the index _role_, there is no issue

```rst
This is normal text, now an equation.

.. math::

   E = m c^2

:index:`Einstein`
:index:`relativity` equation is famous.
And some more text.
```

as it produces

```latex
This is normal text, now an equation.
\begin{equation*}
\begin{split}E = m c^2\end{split}
\end{equation*}
\index{Einstein}Einstein
\index{relativity}relativity equation is famous.
And some more text.
```

It seems to me this reveals a LaTeX issue, but anyway your fix looks good to me, <strike>however I need to make sure that never will latex writer then put things on same line ending with `%`.</strike> (misread the diff, sorry)

**Comment by jfbu** [comment]
However the extra blank line shown in your latex output can induce some vertical space difference after lists. (and even after equation environement it is not good, in LaTeX equations are not really thought out to be paragraphs of their own, they are thought to be mid-paragraph).

edit: an extra blank line can facilitate a pagebrak and then the page number in the index will be off by one. A similar issue #4071 already exists with produced latex file when handling ``option``  directive.

**Comment by jfbu** [comment]
For reference, this problem with `\index` is a pure LaTeX one, see e.g [this question on tex.sx](https://tex.stackexchange.com/questions/26491/index-causing-spurious-whitespace) and [this other one](https://tex.stackexchange.com/questions/60507/index-after-mathmode)

File: sphinx/writers/latex.py
Comment (jfbu) [inline sphinx/writers/latex.py]: This causes blank line when text paragraph starts afterwards. This blank line may facilitate a page break and then the indexed page number will be off by one. For this reason I am considering using

    self.body.append('\ignorespaces ')

Another approach is to insert something like `\ifhmode\hskip-1pt\hskip1pt\relax\fi` before the first `\index`. This is actually extracted from comments in the LaTeX source code! (whow). But its is complicated and its only purpose is to force a non-zero `\lastskip` which triggers the `\index` command to issue the `\ignorespaces`. Thus, maybe easier to output ourselves the `\ignorespaces ` systematically.

**Review by jfbu — CHANGES_REQUESTED** [review]

File: sphinx/writers/latex.py
Comment (jfbu) [reply sphinx/writers/latex.py]: Please read

    self.body.append('\\ignorespaces ')

sorry, got rusty during summer ...

Comment (jfbu) [reply sphinx/writers/latex.py]: The reason we at Sphinx can use `\ignorespaces` is that in inline mode, the `:index:` in rst source must have a space before. Thus there is a space before `\index` in output. Hence the `\ignorespaces` after does not cause two successive words to touch.

Comment (jfbu) [reply sphinx/writers/latex.py]: Hmm... I really had a too long vacation. When using the ````:index:`foo` ```` role, the latex will contain `\index{foo}foo` and there is no space after `\index{foo}` at all anyway. What I wanted to say I guess is that Sphinx never produces `foo\index{bar} bar` type of output currently, so `foo\index{bar}\ignorespaces  bar` which would typeset in PDF `foobar` will never happen if we use `\ignorespaces ` extra.

I think 

    self.body.append('\\ignorespaces ')

is the way to go.

**Comment by aszekMosek** [comment]
Thanks for looking into this! I initially implemented the fix that worked best for me.

**Review by jfbu — APPROVED** [review]

**Comment by jfbu** [comment]
LGTM! can you please rebase to HEAD of stable branch?

File: sphinx/writers/latex.py
Comment (jfbu) [inline sphinx/writers/latex.py]: It could be made conditional

                if not node.get('inline', True):
                    self.body.append('\\ignorespaces ')

to limit a bit the number of `\ignorespaces ` added to latex file.

An alternative would be for Sphinx LaTeX style file to redefine `\index` to always do the `\ignorespaces `, so that the latter would be hidden in redefined `\index` macro. As there is also issue #4071 related to `index`, final decision can be postponed.

**Review by jfbu — APPROVED** [review]

File: sphinx/writers/latex.py
Comment (jfbu) [reply sphinx/writers/latex.py]: ah sorry, I think the

        if not node.get('inline', True):
            self.body.append('\\ignorespaces ')

could be moved out of the `for` loop and inserted right before the final

    raise nodes.SkipNode

this will reduce the number of  `\ignorespaces ` added to latex file. And in case of error with the index key, the added `\ignorespaces ` causes no harm anyhow in latex output.

**Comment by jfbu** [comment]
@tk0miya `AppVeyor was unable to build non-mergeable pull-request` is strange message. I can't access details to check for cause of failure. Maybe appveyor started before the merge-to branch was changed to stable and needs to be re-launched on this ? (I don't know how to do that, perhaps closing and opening again) (edit: only needed to push some new commit...)

@aszekMosek if you don't mind after merging once appveyor is happy I will add commit to implement my comment at https://github.com/sphinx-doc/sphinx/pull/4066#discussion_r139635235 (or if you agree you can push a commit yourself). This will generate less `\ignorespaces`  in latex output. As per the reason why we could not keep the `%\n` initial proposal, I commented more at https://github.com/sphinx-doc/sphinx/pull/4066#issuecomment-330259511: it would create more potential cases of off-by-one page numbers in index, and we already have that problem (see #4071.) And I will add a test.

**Comment by aszekMosek** [comment]
@jfbu Feel free to continue this whatever way fits best, also just close this and make your own changes if that's easier. Thanks.

**Comment by jfbu** [comment]
Thanks for contribution. The bug you reported can actually be seen in current sphinx.pdf (documentation of ``rst_epilog``  in General Configuration section). I will merge when tests complete.

**Pull request merged**
